mlir/lib/Quantum/Transforms/annotate_function.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Quantum/IR/QuantumOps.h"
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+static constexpr const char *hasMeasureAttrName = "catalyst.invalidGradientOperation";
+
+bool isAnnotated(func::FuncOp op, const char *attr)
+{
+    return (bool)(op->getAttrOfType<UnitAttr>(attr));
+}
+
+bool invalidGradientOperation(func::FuncOp op)
+{
+    auto res = op.walk([](Operation *o) {
+        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::PythonCallOp>(o) ||
+            dyn_cast<catalyst::CustomCallOp>(o)) {
+            return WalkResult::interrupt();
+        }
+        else {
+            return WalkResult::advance();
+        }
+    });
+    return res.wasInterrupted();
+}
+
+bool successfulMatchLeaf(func::FuncOp op)
+{
+    return !isAnnotated(op, hasMeasureAttrName) && invalidGradientOperation(op);
+}
+
+void annotate(func::FuncOp op, PatternRewriter &rewriter, const char *attr)
+{
+    op->setAttr(attr, rewriter.getUnitAttr());
+}
+
+struct AnnotateFunctionTransform : public OpRewritePattern<func::FuncOp> {
```

Comment:
Can you call these patterns instead of transforms (in the class name)? Easier to read when sticking with the MLIR coding conventions.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/204/commits/4978a063f00842268cf10f89568e91c59553f663